src/Streamly/Internal/Data/Array/Stream/Fold/Foreign.hs

```diff
@@ -0,0 +1,343 @@
+-- |
+-- Module      : Streamly.Internal.Data.Array.Stream.Fold.Foreign
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Fold a stream of foreign arrays.  @Fold m a b@ in this module works
+-- on a stream of "Array a" and produces an output of type @b@.
+--
+-- Though @Fold m a b@ in this module works on a stream of @Array a@ it is
+-- different from @Data.Fold m (Array a) b@.  While the latter works on arrays
+-- as a whole treating them as atomic elements, the folds in this module can
+-- work on the stream of arrays as if it is an element stream with all the
+-- arrays coalesced together. This module allows adapting the element stream
+-- folds in Data.Fold to correctly work on an array stream as if it is an
+-- element stream. For example:
+--
+-- >>> import qualified Streamly.Data.Fold as Fold
+-- >>> import qualified Streamly.Internal.Data.Array.Stream.Foreign as ArrayStream
+-- >>> import qualified Streamly.Internal.Data.Array.Stream.Fold.Foreign as ArrayFold
+-- >>> import qualified Streamly.Internal.Data.Stream.IsStream as Stream (arraysOf)
+-- >>> import qualified Streamly.Prelude as Stream
+--
+-- >>> ArrayStream.fold (ArrayFold.fromFold (Fold.take 7 Fold.toList)) $ Stream.arraysOf 5 $ Stream.fromList "hello world"
+-- "hello w"
+--
+module Streamly.Internal.Data.Array.Stream.Fold.Foreign
+    (
+      Fold (..)
+
+    -- * Construction
+    , fromFold
+    , fromParser
+    , fromArrayFold
+
+    -- * Mapping
+    , rmapM
+
+    -- * Applicative
+    , yield
+    , yieldM
+    , serialWith
+
+    -- * Monad
+    , concatMap
+
+    -- * Transformation
+    , take
+    )
+where
+
+import Control.Applicative (liftA2)
+import Control.Exception (assert)
+import Control.Monad.Catch (MonadThrow)
+import Control.Monad.IO.Class (MonadIO(..))
+import Foreign.ForeignPtr (touchForeignPtr)
+import Foreign.Ptr (minusPtr, plusPtr)
+import Foreign.Storable (Storable(..))
+import GHC.ForeignPtr (ForeignPtr(..))
+import GHC.Ptr (Ptr(..))
+import GHC.Types (SPEC(..))
+import Streamly.Internal.Data.Array.Foreign.Type (Array(..))
+import Streamly.Internal.Data.Parser.ParserD (Initial(..), Step(..))
+import Streamly.Internal.Data.Tuple.Strict (Tuple'(..))
+
+import qualified Streamly.Internal.Data.Array.Foreign as Array
+import qualified Streamly.Internal.Data.Fold as Fold
+import qualified Streamly.Internal.Data.Parser.ParserD as ParserD
+import qualified Streamly.Internal.Data.Parser.ParserD.Type as ParserD
+
+import Prelude hiding (concatMap, take)
+
+-- | Array stream fold.
+--
+-- An array stream fold is basically an array stream "Parser" that does not
+-- fail.  In case of array stream folds the count in 'Partial', 'Continue' and
+-- 'Done' is a count of elements that includes the leftover element count in
```

Comment:
This can be rephrased.

Reply:
How do you want to rephrase it? I am not focusing on doc yet.

Reply:
I had to read this multiple times to understand. Can be rephrased such that `array stream folds` is read as 1 word.